USAREC_DS_Skills_Assess_20240903_V1.Rmd

```diff
@@ -0,0 +1,221 @@
+
+```{r}
+
+# 3. Get the data ready for the model
+
+# Define the file path for the CSV
+file_path <- 'C:/Users/lysad/OneDrive/Desktop/Viper/USAREC DS Technical Assessment/Code Output/sql_applicants.csv'
+
+# Function to read and prepare data
+prepare_data <- function(file_path) {
+  applicants_data <- readr::read_csv(file_path)
+  
+  # Data preprocessing
+  applicants_data <- applicants_data %>%
+    mutate(fymo = paste(fiscal_year, fiscal_month, sep = '-'),
+           unit_sublevel_1 = str_sub(unit_id, 1, 1),
+           unit_sublevel_2 = str_sub(unit_id, 2, 1)) %>%
+    select(-fiscal_year, -fiscal_month)
+  
+  return(applicants_data)
+}
+
+# Prepare the data
+applicants_data <- prepare_data(file_path)
+
+# Summary of the data
+summary(applicants_data)
+
+# Detailed descriptive statistics
+skim(applicants_data)
+```
+
+```{r}
+# 4. Fit the model and write the coefficients
+
+fit_and_save_model <- function(data, output_file) 
+  {
+  # Fit the model
+  model <- lm(sum_of_applicants ~ ., data = data)
+  
+  # Plot the model
+  plot(model)
+  
+  # Extract coefficients
+  beta_fits <- coef(model)
+  
+  # Save coefficients to CSV
+  write_csv(data.frame(beta_fits), output_file)
+  
+  return(list(model = model, beta_fits = beta_fits))
+ }
+
+# Model output file path
+output_file <- 'C:/Users/lysad/OneDrive/Desktop/Viper/USAREC DS Technical Assessment/Code Output/applicants_beta_fitted.csv'
+
+# Fit the model and generate coefficients
+model_results <- fit_and_save_model(applicants_data, output_file)
+
+# Extract model and beta_fits from results
+model <- model_results$model
+beta_fits <- model_results$beta_fits
+
+```
+
+```{r}
+
+# Convert to a heatmap correlation matrix of coefficients 
+coeff_matrix <- matrix(beta_fits, 
+                       nrow = 1, 
+                       dimnames = list("Coefficients", names(beta_fits)))
+
+# Create heatmap
+pheatmap(coeff_matrix, 
+         cluster_rows = FALSE, 
+         cluster_cols = FALSE, 
+         display_numbers = TRUE)
+
+```
+
+```{r}
+
+# Perform ANOVA
+anova_results <- anova(model)
+
+# Print ANOVA results
+print(anova_results)
+
+```
+
+
+```{r}
+
+# 5. Cross-validation
+
+validate_model <- function(data) 
+  {
+  train_control <- trainControl(method = "cv", 
+                                number = 10)
+
+  cv_applicants_model <- train(sum_of_applicants ~ ., 
+                    data = data, method = "lm", 
+                    trControl = train_control)
+  
+  return(cv_applicants_model)
+ }
+
+# Perform cross-validation
+cv_applicants_model_results <- validate_model(applicants_data)
+
+# Print cross-validation results
+print(cv_applicants_model_results)
+
+# Save cross-validation results to CSV
+cv_applicants_model_results <- as.data.frame(cv_applicants_model_results$results)
+
+output_cv_file <- 'C:/Users/lysad/OneDrive/Desktop/Viper/USAREC DS Technical Assessment/Code Output/cv_applicants_model_results.csv'
+
+write_csv(cv_applicants_model_results, output_cv_file)
+
+#End
+
```

Comment:
Lydia, thank you for taking the time to provide a well thought-out submission. In looking at your changes you clearly have a great grasp of an ideal DS workflow and a thorough understanding of what the overall project is accomplishing if it were a real project. A couple of pointers that could possibly help are:
  1. Consider a more modular approach to the code by using multiple `.R` scripts instead of one very large `.Rmd`. This makes it easier to test and maintain if I only have to look at a few files of a few lines each as opposed to one file of 215 lines. A key difference that we are looking for between data engineering and data science is the candidate's ability to use best software practices.
  2. In your write-up, which showcases how well you do understand the DS workflow, again less may have been more here as I was hoping to just understand the changes you made and not necessarily the theoretical underpinnings.